{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC_number</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>_seq</th>\n",
       "      <th>annotations</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>description</th>\n",
       "      <th>experiment</th>\n",
       "      <th>gene</th>\n",
       "      <th>...</th>\n",
       "      <th>organism</th>\n",
       "      <th>product</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>pseudo</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>strain</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>translation</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[taxon:93061]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[Staphylococcus aureus subsp. aureus NCTC 8325]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[NCTC 8325]</td>\n",
       "      <td>[aureus]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>source</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GeneID:3919798]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[dnaA]</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[GeneID:3919798]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[dnaA]</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[chromosomal replication initiation protein]</td>\n",
       "      <td>[YP_498609.1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GeneID:3919799]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[2.7.7.7]</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[GeneID:3919799]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[DNA polymerase III subunit beta]</td>\n",
       "      <td>[YP_498610.1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVIL...</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   EC_number _per_letter_annotations  \\\n",
       "0        NaN                      {}   \n",
       "1        NaN                      {}   \n",
       "2        NaN                      {}   \n",
       "3        NaN                      {}   \n",
       "4  [2.7.7.7]                      {}   \n",
       "\n",
       "                                                _seq  \\\n",
       "0  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "1  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "2  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "3  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "4  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "\n",
       "                                         annotations codon_start  \\\n",
       "0  {'molecule_type': 'DNA', 'topology': 'circular...         NaN   \n",
       "1  {'molecule_type': 'DNA', 'topology': 'circular...         NaN   \n",
       "2  {'molecule_type': 'DNA', 'topology': 'circular...         [1]   \n",
       "3  {'molecule_type': 'DNA', 'topology': 'circular...         NaN   \n",
       "4  {'molecule_type': 'DNA', 'topology': 'circular...         [1]   \n",
       "\n",
       "            db_xref                                            dbxrefs  \\\n",
       "0     [taxon:93061]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "1  [GeneID:3919798]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "2  [GeneID:3919798]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "3  [GeneID:3919799]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "4  [GeneID:3919799]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "\n",
       "                                         description experiment    gene  ...  \\\n",
       "0  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN     NaN  ...   \n",
       "1  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN  [dnaA]  ...   \n",
       "2  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN  [dnaA]  ...   \n",
       "3  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN     NaN  ...   \n",
       "4  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN     NaN  ...   \n",
       "\n",
       "                                          organism  \\\n",
       "0  [Staphylococcus aureus subsp. aureus NCTC 8325]   \n",
       "1                                              NaN   \n",
       "2                                              NaN   \n",
       "3                                              NaN   \n",
       "4                                              NaN   \n",
       "\n",
       "                                        product     protein_id pseudo  \\\n",
       "0                                           NaN            NaN    NaN   \n",
       "1                                           NaN            NaN    NaN   \n",
       "2  [chromosomal replication initiation protein]  [YP_498609.1]    NaN   \n",
       "3                                           NaN            NaN    NaN   \n",
       "4             [DNA polymerase III subunit beta]  [YP_498610.1]    NaN   \n",
       "\n",
       "  pseudogene       strain sub_species transl_table  \\\n",
       "0        NaN  [NCTC 8325]    [aureus]          NaN   \n",
       "1        NaN          NaN         NaN          NaN   \n",
       "2        NaN          NaN         NaN         [11]   \n",
       "3        NaN          NaN         NaN          NaN   \n",
       "4        NaN          NaN         NaN         [11]   \n",
       "\n",
       "                                         translation    type  \n",
       "0                                                NaN  source  \n",
       "1                                                NaN    gene  \n",
       "2  [MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...     CDS  \n",
       "3                                                NaN    gene  \n",
       "4  [MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVIL...     CDS  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import Bio\n",
    "import pandas\n",
    "from pandas import DataFrame\n",
    "import numpy as np\n",
    "from pandas.core.frame import *\n",
    "from types import GeneratorType\n",
    "from typing import Union, Dict, List, Tuple\n",
    "from Bio import SeqIO\n",
    "from copy import deepcopy\n",
    "import gzip\n",
    "from pathlib import Path\n",
    "from Bio.Alphabet import single_letter_alphabet\n",
    "\n",
    "\n",
    "class SubclassedSeries(pandas.Series):\n",
    "    \"\"\" Pandas Series API to Inherit \"\"\"\n",
    "    \n",
    "    @property\n",
    "    def _constructor(self):\n",
    "        return SubclassedSeries\n",
    "\n",
    "    @property\n",
    "    def _constructor_expanddim(self):\n",
    "        return SubclassedDataFrame\n",
    "\n",
    "\n",
    "class SubclassedDataFrame(pandas.DataFrame):\n",
    "    \"\"\" Pandas DataFrame to Inherit \"\"\"\n",
    "    \n",
    "    @property\n",
    "    def _constructor(self):\n",
    "        return SubclassedDataFrame\n",
    "\n",
    "    @property\n",
    "    def _constructor_sliced(self):\n",
    "        return SubclassedSeries\n",
    "\n",
    "\n",
    "class BioDatabase(SubclassedDataFrame):\n",
    "    \"\"\" Expanded Pandas DataFrame to handle BioPython SeqRecords generator or genomic file types \"\"\"\n",
    "    \n",
    "    @classmethod\n",
    "    def from_seqrecords(cls, seqrecords: Union[GeneratorType, list], index=None, exclude=None, \n",
    "                        columns=None, coerce_float=False, nrows=None) -> pandas.DataFrame:\n",
    "        \"\"\" Takes Biopython parsed output to convert to a proper DataFrame \n",
    "\n",
    "        :param seqrecords: Generator or list from BioPython universal output. \n",
    "            All formats are the same output.\n",
    "        \n",
    "        >>> from_seqrecords(Bio.SeqIO.parse('file.fasta', format='fasta'))\n",
    "        \"\"\"\n",
    "        # if isinstance(seqrecords, GeneratorType):\n",
    "        data = cls.__normalize_seqrecords(seqrecords)\n",
    "        # else:\n",
    "        #     data = seqrecords\n",
    "        return cls.from_records(data, index=index, exclude=exclude, columns=columns,\n",
    "                                coerce_float=coerce_float, nrows=nrows)\n",
    "\n",
    "    def __normalize_seqrecords(seqrecords: Union[GeneratorType, list]) -> List[dict]:\n",
    "        \"\"\" Pull nested dictionaries into a single dictionary.\n",
    "\n",
    "        Priority is given to the keys higher in the hierarchy.\n",
    "        :param seqrecords: Generator from BioPython universal output. \n",
    "            All formats are the same output.\n",
    "        :returns: List of dictionaries with keys that were obtained along the way.\n",
    "        \n",
    "        >>> __normalize_seqrecords(Bio.SeqIO.parse('file.fasta', format='fasta'))\n",
    "        \"\"\"\n",
    "        records = []\n",
    "        for seqrecord in SeqIO.to_dict(seqrecords).values():\n",
    "            _records = []\n",
    "            record = seqrecord.__dict__\n",
    "            # If a more complicated format is used; features will be nested.\n",
    "            features = record.pop('features') if record.get('features') else []\n",
    "            for feature in features:\n",
    "                _record = deepcopy(record)\n",
    "                # Meta that make up the feature\n",
    "                aspects = feature.__dict__\n",
    "                # Qualifier dictionary inside each feature\n",
    "                qualifiers = aspects.pop('qualifiers') if aspects.get('qualifiers') else {}\n",
    "                # Add each feature aspect\n",
    "                for aspect_key, aspect_value in aspects.items():\n",
    "                    if aspect_key not in record:\n",
    "                        _record[aspect_key] = aspect_value\n",
    "                # Add each qualifier\n",
    "                for qualifier_key, qualifier_value in qualifiers.items():\n",
    "                    _record = deepcopy(_record)\n",
    "                    if qualifier_key not in _record:\n",
    "                        _record[qualifier_key] = qualifier_value\n",
    "                # Collect normalized feature \n",
    "                _records += [_record]\n",
    "            # If no normalized feature collected use original seq record\n",
    "            if not _records:\n",
    "                _records += [record]\n",
    "            # Add current records list to past iterations.\n",
    "            # We do this because there could be more than one feature per seqrecord.\n",
    "            records += _records\n",
    "        return records\n",
    "\n",
    "\n",
    "def pathing(path: Union[str, Path], new: bool = False) -> Path:\n",
    "    \"\"\" Guarantees correct expansion rules for pathing.\n",
    "\n",
    "    :param Union[str, Path] path: path of folder or file you wish to expand.\n",
    "    :param bool new: will check if distination exists if new\n",
    "        (will check parent path regardless).\n",
    "    :return: A pathlib.Path object.\n",
    "    \n",
    "    >>> pathing('~/Desktop/folderofgoodstuffs/')\n",
    "    /home/user/Desktop/folderofgoodstuffs\n",
    "    \"\"\"\n",
    "    path = Path(path)\n",
    "    # Expand tilda shortened path or local path.\n",
    "    if str(path)[0] == '~':\n",
    "        path = path.expanduser()\n",
    "    else:\n",
    "        path = path.absolute() #\n",
    "    # Making sure new paths don't exist while also making sure existing paths actually exist.\n",
    "    if new:\n",
    "        if not path.parent.exists():\n",
    "            raise ValueError(f'ERROR ::: Parent directory of {path} does not exist.')\n",
    "        if path.exists():\n",
    "            raise ValueError(f'ERROR ::: {path} already exists!')\n",
    "    else:\n",
    "        if not path.exists():\n",
    "            raise ValueError(f'ERROR ::: Path {path} does not exist.')\n",
    "    return path\n",
    "\n",
    "\n",
    "def read_seq(handle: str, format: str, alphabet: object = single_letter_alphabet) -> pandas.DataFrame:\n",
    "    \"\"\" Read Bioinformatic file type \n",
    "    \n",
    "    :param str handle: str path of file to open.\n",
    "    :param str format: Broad range of Bioinformatic formats ie fasta & genbank.\n",
    "    :param object alphabet: Custom string from BioPython with handy methods.\n",
    "    \n",
    "    >>> read_seq('file.fasta.gz', format='fasta')\n",
    "    \"\"\"\n",
    "    # Checks path validity\n",
    "    path = pathing(handle)\n",
    "    # If file is gzip compressed\n",
    "    if path.suffix == '.gz':\n",
    "        with gzip.open(handle, \"rt\") as handle:\n",
    "            seqrecords = SeqIO.parse(handle, format=format, alphabet=alphabet)\n",
    "            # need to use/return while I/O is open\n",
    "            return BioDatabase.from_seqrecords(seqrecords)\n",
    "    # Uncompressed; will break if another compression is used.\n",
    "    seqrecords = SeqIO.parse(handle, format=format, alphabet=alphabet)\n",
    "    return BioDatabase.from_seqrecords(seqrecords)\n",
    "\n",
    "\n",
    "# Add to pandas module for seemly behavior\n",
    "pandas.DataFrame = BioDatabase\n",
    "pandas.read_seq = read_seq\n",
    "\n",
    "pd = pandas\n",
    "df = pandas.read_seq(gbk, format='genbank')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample files used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "gbk = 'GCF_000013425.1.gbk.gz'\n",
    "fasta = 'random_sequences.fasta.gz'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fasta to DF example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>_seq</th>\n",
       "      <th>annotations</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>description</th>\n",
       "      <th>features</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{}</td>\n",
       "      <td>(G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  _per_letter_annotations                                               _seq  \\\n",
       "0                      {}  (C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...   \n",
       "1                      {}  (G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...   \n",
       "2                      {}  (C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...   \n",
       "\n",
       "  annotations dbxrefs                                        description  \\\n",
       "0          {}      []  FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...   \n",
       "1          {}      []  FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...   \n",
       "2          {}      []  FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...   \n",
       "\n",
       "  features              id            name  \n",
       "0       []  FWIRNKE01DKIF6  FWIRNKE01DKIF6  \n",
       "1       []  FWIRNKE01CDBE3  FWIRNKE01CDBE3  \n",
       "2       []  FWIRNKE01BKZJJ  FWIRNKE01BKZJJ  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_seq(fasta, format='fasta')\n",
    "df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Search description with little effort!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>_seq</th>\n",
       "      <th>annotations</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>description</th>\n",
       "      <th>features</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  _per_letter_annotations                                               _seq  \\\n",
       "0                      {}  (C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...   \n",
       "\n",
       "  annotations dbxrefs                                        description  \\\n",
       "0          {}      []  FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...   \n",
       "\n",
       "  features              id            name  \n",
       "0       []  FWIRNKE01DKIF6  FWIRNKE01DKIF6  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I want all ranks of a certain number \n",
    "df[df['description'].str.contains('rank=0000177')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GenBank to DF example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC_number</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>_seq</th>\n",
       "      <th>annotations</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>description</th>\n",
       "      <th>experiment</th>\n",
       "      <th>gene</th>\n",
       "      <th>...</th>\n",
       "      <th>organism</th>\n",
       "      <th>product</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>pseudo</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>strain</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>translation</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[taxon:93061]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[Staphylococcus aureus subsp. aureus NCTC 8325]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[NCTC 8325]</td>\n",
       "      <td>[aureus]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>source</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GeneID:3919798]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[dnaA]</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[GeneID:3919798]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[dnaA]</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[chromosomal replication initiation protein]</td>\n",
       "      <td>[YP_498609.1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  EC_number _per_letter_annotations  \\\n",
       "0       NaN                      {}   \n",
       "1       NaN                      {}   \n",
       "2       NaN                      {}   \n",
       "\n",
       "                                                _seq  \\\n",
       "0  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "1  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "2  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "\n",
       "                                         annotations codon_start  \\\n",
       "0  {'molecule_type': 'DNA', 'topology': 'circular...         NaN   \n",
       "1  {'molecule_type': 'DNA', 'topology': 'circular...         NaN   \n",
       "2  {'molecule_type': 'DNA', 'topology': 'circular...         [1]   \n",
       "\n",
       "            db_xref                                            dbxrefs  \\\n",
       "0     [taxon:93061]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "1  [GeneID:3919798]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "2  [GeneID:3919798]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "\n",
       "                                         description experiment    gene  ...  \\\n",
       "0  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN     NaN  ...   \n",
       "1  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN  [dnaA]  ...   \n",
       "2  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN  [dnaA]  ...   \n",
       "\n",
       "                                          organism  \\\n",
       "0  [Staphylococcus aureus subsp. aureus NCTC 8325]   \n",
       "1                                              NaN   \n",
       "2                                              NaN   \n",
       "\n",
       "                                        product     protein_id pseudo  \\\n",
       "0                                           NaN            NaN    NaN   \n",
       "1                                           NaN            NaN    NaN   \n",
       "2  [chromosomal replication initiation protein]  [YP_498609.1]    NaN   \n",
       "\n",
       "  pseudogene       strain sub_species transl_table  \\\n",
       "0        NaN  [NCTC 8325]    [aureus]          NaN   \n",
       "1        NaN          NaN         NaN          NaN   \n",
       "2        NaN          NaN         NaN         [11]   \n",
       "\n",
       "                                         translation    type  \n",
       "0                                                NaN  source  \n",
       "1                                                NaN    gene  \n",
       "2  [MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...     CDS  \n",
       "\n",
       "[3 rows x 27 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gbk_df = pd.read_seq(gbk, format='genbank')\n",
    "gbk_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# See all the possible columns of meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['EC_number', '_per_letter_annotations', '_seq', 'annotations',\n",
       "       'codon_start', 'db_xref', 'dbxrefs', 'description', 'experiment',\n",
       "       'gene', 'gene_synonym', 'id', 'location', 'locus_tag', 'mol_type',\n",
       "       'name', 'note', 'organism', 'product', 'protein_id', 'pseudo',\n",
       "       'pseudogene', 'strain', 'sub_species', 'transl_table', 'translation',\n",
       "       'type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gbk_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Seach for only rows of type CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome has CDS count: (2767, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC_number</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>_seq</th>\n",
       "      <th>annotations</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>description</th>\n",
       "      <th>experiment</th>\n",
       "      <th>gene</th>\n",
       "      <th>...</th>\n",
       "      <th>organism</th>\n",
       "      <th>product</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>pseudo</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>strain</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>translation</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[GeneID:3919798]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[dnaA]</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[chromosomal replication initiation protein]</td>\n",
       "      <td>[YP_498609.1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[2.7.7.7]</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[GeneID:3919799]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[DNA polymerase III subunit beta]</td>\n",
       "      <td>[YP_498610.1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVIL...</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>{}</td>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>{'molecule_type': 'DNA', 'topology': 'circular...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[GeneID:3919176]</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[hypothetical protein]</td>\n",
       "      <td>[YP_498611.1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[MIILVQEVVVEGDINLGQFLKTEGIIESGGQAKWFLQDVEVLING...</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   EC_number _per_letter_annotations  \\\n",
       "2        NaN                      {}   \n",
       "4  [2.7.7.7]                      {}   \n",
       "6        NaN                      {}   \n",
       "\n",
       "                                                _seq  \\\n",
       "2  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "4  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "6  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...   \n",
       "\n",
       "                                         annotations codon_start  \\\n",
       "2  {'molecule_type': 'DNA', 'topology': 'circular...         [1]   \n",
       "4  {'molecule_type': 'DNA', 'topology': 'circular...         [1]   \n",
       "6  {'molecule_type': 'DNA', 'topology': 'circular...         [1]   \n",
       "\n",
       "            db_xref                                            dbxrefs  \\\n",
       "2  [GeneID:3919798]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "4  [GeneID:3919799]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "6  [GeneID:3919176]  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]   \n",
       "\n",
       "                                         description experiment    gene  ...  \\\n",
       "2  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN  [dnaA]  ...   \n",
       "4  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN     NaN  ...   \n",
       "6  Staphylococcus aureus subsp. aureus NCTC 8325 ...        NaN     NaN  ...   \n",
       "\n",
       "  organism                                       product     protein_id  \\\n",
       "2      NaN  [chromosomal replication initiation protein]  [YP_498609.1]   \n",
       "4      NaN             [DNA polymerase III subunit beta]  [YP_498610.1]   \n",
       "6      NaN                        [hypothetical protein]  [YP_498611.1]   \n",
       "\n",
       "  pseudo pseudogene strain sub_species transl_table  \\\n",
       "2    NaN        NaN    NaN         NaN         [11]   \n",
       "4    NaN        NaN    NaN         NaN         [11]   \n",
       "6    NaN        NaN    NaN         NaN         [11]   \n",
       "\n",
       "                                         translation type  \n",
       "2  [MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...  CDS  \n",
       "4  [MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVIL...  CDS  \n",
       "6  [MIILVQEVVVEGDINLGQFLKTEGIIESGGQAKWFLQDVEVLING...  CDS  \n",
       "\n",
       "[3 rows x 27 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds = gbk_df[gbk_df.type == 'CDS']\n",
    "print('Genome has CDS count:', cds.shape)\n",
    "cds.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Better view of the layout for a single row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'EC_number': {2: nan},\n",
       " '_per_letter_annotations': {2: {}},\n",
       " '_seq': {2: Seq('CGATTAAAGATAGAAATACACGATGCGAGCAATCAAATTTCATAACATCACCAT...TAT', SingleLetterAlphabet())},\n",
       " 'annotations': {2: {'molecule_type': 'DNA',\n",
       "   'topology': 'circular',\n",
       "   'data_file_division': 'CON',\n",
       "   'date': '03-AUG-2016',\n",
       "   'accessions': ['NC_007795'],\n",
       "   'sequence_version': 1,\n",
       "   'keywords': ['RefSeq'],\n",
       "   'source': 'Staphylococcus aureus subsp. aureus NCTC 8325',\n",
       "   'organism': 'Staphylococcus aureus subsp. aureus NCTC 8325',\n",
       "   'taxonomy': ['Bacteria',\n",
       "    'Firmicutes',\n",
       "    'Bacilli',\n",
       "    'Bacillales',\n",
       "    'Staphylococcaceae',\n",
       "    'Staphylococcus'],\n",
       "   'references': [Reference(title='The Staphylococcus aureus NCTC8325 Genome', ...),\n",
       "    Reference(title='Direct Submission', ...),\n",
       "    Reference(title='Direct Submission', ...)],\n",
       "   'comment': 'REVIEWED REFSEQ: This record has been curated by NCBI staff. The\\nreference sequence was derived from CP000253.\\nRefSeq Category: Reference Genome\\n            UPR: UniProt Genome\\nStaphylococcus aureus subsp. aureus NCTC 8325 is available from\\nwww.narsa.net.\\nCOMPLETENESS: full length.',\n",
       "   'contig': 'join(CP000253.1:1..2821361)'}},\n",
       " 'codon_start': {2: ['1']},\n",
       " 'db_xref': {2: ['GeneID:3919798']},\n",
       " 'dbxrefs': {2: ['BioProject:PRJNA57795', 'Assembly:GCF_000013425.1']},\n",
       " 'description': {2: 'Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome'},\n",
       " 'experiment': {2: nan},\n",
       " 'gene': {2: ['dnaA']},\n",
       " 'gene_synonym': {2: nan},\n",
       " 'id': {2: 'NC_007795.1'},\n",
       " 'location': {2: FeatureLocation(ExactPosition(516), ExactPosition(1878), strand=1)},\n",
       " 'locus_tag': {2: ['SAOUHSC_00001']},\n",
       " 'mol_type': {2: nan},\n",
       " 'name': {2: 'NC_007795'},\n",
       " 'note': {2: ['binds to the dnaA-box as an ATP-bound complex at the origin of replication during the initiation of chromosomal replication; can also affect transcription of multiple genes including itself.']},\n",
       " 'organism': {2: nan},\n",
       " 'product': {2: ['chromosomal replication initiation protein']},\n",
       " 'protein_id': {2: ['YP_498609.1']},\n",
       " 'pseudo': {2: nan},\n",
       " 'pseudogene': {2: nan},\n",
       " 'strain': {2: nan},\n",
       " 'sub_species': {2: nan},\n",
       " 'transl_table': {2: ['11']},\n",
       " 'translation': {2: ['MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSSIPFNANWLNQQYAEIIQAILFDVVGYEVKPHFITTEELANYSNNETATPKETTKPSTETTEDNHVLGREQFNAHNTFDTFVIGPGNRFPHAASLAVAEAPAKAYNPLFIYGGVGLGKTHLMHAIGHHVLDNNPDAKVIYTSSEKFTNEFIKSIRDNEGEAFRERYRNIDVLLIDDIQFIQNKVQTQEEFFYTFNELHQNNKQIVISSDRPPKEIAQLEDRLRSRFEWGLIVDITPPDYETRMAILQKKIEEEKLDIPPEALNYIANQIQSNIRELEGALTRLLAYSQLLGKPITTELTAEALKDIIQAPKSKKITIQDIQKIVGQYYNVRIEDFSAKKRTKSIAYPRQIAMYLSRELTDFSLPKIGEEFGGRDHTTVIHAHEKISKDLKEDPIFKQEVENLEKEIRNV']},\n",
       " 'type': {2: 'CDS'}}"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds.head(1).to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extra Notes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating a BioPython seq from scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGCATGATGATGATGATGATAG', SingleLetterAlphabet())"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = Seq('ATGCATGATGATGATGATGATAG', alphabet=single_letter_alphabet)\n",
    "seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## You can create a dataframe row as such"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    (A, T, G, C, A, T, G, A, T, G, A, T, G, A, T, ...\n",
       "Name: _seq, dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame([{'_seq': seq}])._seq "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
